[Merged by Bors] - 0.8 Migration guide

IceSentry opened the pull request:
## Objective

- Migration guide for 0.8

## Solution

- Create a new section for each PR tagged with `C-Breaking-Change`
	- Copy the Migration Guide from the PR if it exists
- Create a few more sections for PR not tagged with the label

## Notes

The ordering was pretty much whatever order github gave me when I searched breaking changes, so if you feel like something should be close to the top please notify me. For example, I moved the camera driven stuff at the top since it's probably going to be the biggest one

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,300 @@
+
+### [Remove .system()](https://github.com/bevyengine/bevy/pull/4499)
+
+You can no longer use `.system()`. It was deprecated in 0.7.0. You can just remove the method call.
```
Should mention assert_is_system for tests.

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,300 @@
+
+### [Mark mutable APIs under ECS storage as pub(crate)](https://github.com/bevyengine/bevy/pull/5065)
+
+<!-- TODO not sure what to say here -->
```
If any user cared, they should file an issue explaining what on earth they were doing.

hymm commented:
task pools changing to a global is probably a breaking change, but PR wasn't labeled as such  
https://github.com/bevyengine/bevy/pull/2250

alice-i-cecile commented:
Helping sort out confusion caused by changes to camera bundles from https://github.com/bevyengine/bevy/pull/4745. This deserves top billing in terms of this version's migration guide.

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,302 @@
+
+<!-- Dear God, I hope not. -->
+
+If you experienced any problems caused by this change, please [create an issue](https://github.com/bevyengine/bevy/issues) explaining _in details_ what you were doing with those apis.
```
```suggestion
If you experienced any problems caused by this change, please [create an issue](https://github.com/bevyengine/bevy/issues) explaining _in detail_ what you were doing with those apis.
```

nicopap commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -0,0 +1,302 @@
+
+UI no longer requires a dedicated camera. `UiCameraBundle` has been removed. `Camera2dBundle` and `Camera3dBundle` now both default to rendering UI as part of their own render graphs. To disable UI rendering for a camera, disable it using the CameraUi component:
+
+```rust
+commands
+    .spawn_bundle(Camera3dBundle::default())
+    .insert(CameraUi {
+        is_enabled: false,
+        ..default()
+    })
+```
````
This needs to be updated

https://github.com/bevyengine/bevy/pull/5234

IceSentry commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -300,3 +301,20 @@ Exhaustive matches on `RenderGraphRunnerError` will need to add a branch to hand
+Thread pools don't need to be stored in a resource anymore since they are now stored globally. You can now use `get()` to access it.
+
+```rust
+// 0.7
+fn spawn_tasks(mut commands: Commands, thread_pool: Res<AsyncComputeTaskPool>) {
+    // Do something with thread_pool
+}
+
+// 0.8
+fn spawn_tasks(mut commands: Commands) {
+    let thread_pool = AsyncComputeTaskPool::get();
+    // Do something with thread_pool
+}
````
Nope, I copy pasted the example using it from the PR and forgot to remove it 😅 Good catch!

nicopap commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -300,3 +301,99 @@ Exhaustive matches on `RenderGraphRunnerError` will need to add a branch to hand
+
+Updated [struct_trait](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates/bevy_reflect/src/struct_trait.rs#L455-L457), [tuple_struct](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates/bevy_reflect/src/tuple_struct.rs#L366-L368), [tuple](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates ...
+
+### [Rename CameraUi](https://github.com/bevyengine/bevy/pull/5234)
+
+Rename CameraUi to UiCameraConfig
```
#5234 was to make clear that 0.8 `UiCameraConfig` **is not** the same as the 0.7 `CameraUi`, so that users migrating do not get confused. This reintroduces this confusion. The PR should simply not be mentioned in the migration guide.

IceSentry replied:
Ah, right, I misunderstood this a bit, makes sense. Can you make a suggestion with those changes? That way github will list you as a co-author.

nicopap replied:
sure thing, see https://github.com/bevyengine/bevy-website/pull/389/files#r922360048

alice-i-cecile replied:
Remember to fix missing code formatting :)

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+### [Move Rect to bevy_ui and rename it to UiRect](https://github.com/bevyengine/bevy/pull/4276)
+
+The `Rect` type got renamed to `UiRect`. To migrate you just have to change every occurrence of `Rect` to `UiRect`.
```
```suggestion
The `Rect` type has been renamed to `UiRect`.
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+### [Rename ElementState to ButtonState](https://github.com/bevyengine/bevy/pull/4314)
+
+The `ElementState` type received a rename and is now called `ButtonState`. To migrate you just have to change every occurrence of `ElementState` to `ButtonState`.
```
```suggestion
The `ElementState` type has been renamed to `ButtonState`.
```

nicopap commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+### [Make reflect_partial_eq return more accurate results](https://github.com/bevyengine/bevy/pull/5210)
+
+Updated [struct_trait](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates/bevy_reflect/src/struct_trait.rs#L455-L457), [tuple_struct](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates/bevy_reflect/src/tuple_struct.rs#L366-L368), [tuple](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates ...
+
+### [Rename CameraUi](https://github.com/bevyengine/bevy/pull/5234)
+
+Rename CameraUi to UiCameraConfig
```
```suggestion
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+### [Improve docs and naming for RawWindowHandle functionality](https://github.com/bevyengine/bevy/pull/4335)
+
+rename `HasRawWindowHandleWrapper` to `ThreadLockedRawWindowHandleWrapper`
```
```suggestion
Renamed `HasRawWindowHandleWrapper` to `ThreadLockedRawWindowHandleWrapper`.
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+rename `HasRawWindowHandleWrapper` to `ThreadLockedRawWindowHandleWrapper`
+
+### [Migrate to encase from crevice](https://github.com/bevyengine/bevy/pull/4339)
```
`code_blocks` are missing for methods here.

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+### [Change gamepad.rs tuples to normal structs](https://github.com/bevyengine/bevy/pull/4519)
+
+The `Gamepad`, `GamepadButton`, `GamepadAxis`, `GamepadEvent` and `GamepadEventRaw` types are now normal structs instead of tuple structs and have a new() function. To migrate change every instantiation to use the `new()` function instead and use the appropriate field names instead of .0 and .1.
```
```suggestion
The `Gamepad`, `GamepadButton`, `GamepadAxis`, `GamepadEvent` and `GamepadEventRaw` types are now normal structs instead of tuple structs and have a new() function.
To migrate change every instantiation to use the `new()` function instead and use the appropriate field names instead of .0 and .1.
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+### [Replace ReadOnlyFetch with ReadOnlyWorldQuery](https://github.com/bevyengine/bevy/pull/4626)
+
+The trait ReadOnlyFetch has been replaced with ReadOnlyWorldQuery along with the WorldQueryGats::ReadOnlyFetch assoc type which has been replaced with <WorldQuery::ReadOnly as WorldQueryGats>::Fetch
```
Code blocks.

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+<!-- TODO make this a bit friendlier -->
+
+If you are now getting query conflicts from `Or`/`AnyOf`/`Option` rip to you and ur welcome for it now being caught.
```
```suggestion
Query conflicts from `Or`/`AnyOf`/`Option` have been made fixed, and made stricter to avoid UB.
If you have new query conflicts due to this you must refactor your systems; consider using `ParamSet`.
```

IceSentry replied:
Did you mean `have been fixed`? Also, I think I'd prefer not using acronyms for UB since this could be read by beginners.

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -0,0 +1,399 @@
+}
+```
+
+If using `Query(State)` outside of a system run by the scheduler, you may need to manually configure and initialize a `ComputeTaskPool` as a resource in the `World`.
````
```suggestion
If using `Query` or `QueryState` outside of a system run by the scheduler, you may need to manually configure and initialize a `ComputeTaskPool` as a resource in the `World`.
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+
+### [Fail to compile on 16-bit platforms](https://github.com/bevyengine/bevy/pull/4736)
+
+`bevy_ecs` will now explicitly fail to compile on 16-bit platforms. If this is required, there is currently no alternative. Please file an issue (<https://github.com/bevyengine/bevy/issues>) to help detail your use case.
```
```suggestion
`bevy_ecs` will now explicitly fail to compile on 16-bit platforms, as the behavior was undefined.

There is currently no alternative, but we're open to adding support.
Please file an issue (<https://github.com/bevyengine/bevy/issues>) to help detail your use case.
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,399 @@
+* a `HandleUntyped`, use `&handle_untyped.typed_weak()` to create a weak handle of the specified type. This is most likely to be the useful when using [load_folder](https://docs.rs/bevy_asset/latest/bevy_asset/struct.AssetServer.html#method.load_folder)
+* a `Handle<U>` of  of a different type, consider whether this is the correct handle type to store. If it is (i.e. the same handle id is used for multiple different Asset types) use `Handle::weak(handle.id)` to cast to a different type.
+
+### [Allow higher order systems](https://github.com/bevyengine/bevy/pull/4833)
```
Ping @DJMcNab. I can't see any breaking behavior in a cursory pass.

DJMcNab replied:
It was *technically* a breaking change, since it changes the public API of `SystemParamFunction`.
I was expecting pushback against it being treated as 'not a breaking change', so I marked it as such to be correct.

Maybe a note that `SystemParamFunction` has been changed, in order to allow composing systems? I still think that there have been no external users, so we can probably just remove this section.

hymm commented:
I think this was a breaking change, but not sure about the details: https://github.com/bevyengine/bevy/pull/4197

rparrett commented:
Users might find their stuff unexpectedly not visible if the parents don't have `Visibility` and `ComputedVisibility` components. We should probably mentioned this in the visibility section.

See https://github.com/bevyengine/bevy/issues/5334

IceSentry commented:
Reminder, need to add a section about https://github.com/bevyengine/bevy/pull/5344

rparrett commented:
`spawn_scene` was removed in favor of `SceneBundle`

https://github.com/bevyengine/bevy/pull/2424

nicopap commented:
Let's see:

* `world_to_screen` ⇒ `world_to_viewport` I think this was part of https://github.com/bevyengine/bevy/pull/4745
* The `Parent` and `Children` component fields are now private. (https://github.com/bevyengine/bevy/pull/4197)
  * replace `parent.0` by `parent.get()`. With `Children`, doing `*children` is probably enough to replace `children.0`
  * You can't construct `Children` or `Parent` component anymore, see [this gist](https://gist.github.com/nicopap/38bc2d648ab5926af41d2b33418fc472) for a stopgap measure, which may introduce a single frame delay
* `GlobalTransform` fields have changed (https://github.com/bevyengine/bevy/pull/4379)
  * Replace `global_transform.translation` by `global_transform.translation()` (For other fields, use the `compute_transform` method)
  * `GlobalTransform` do not support non-linear scales anymore, we'd like to hear from you if it is an inconvenience for you

DJMcNab commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,401 @@
+
+### [Make paused timers update just_finished on tick](https://github.com/bevyengine/bevy/pull/4445)
+
+`Timer::times_finished has` been renamed to `Timer::times_finished_this_tick` for clarity.
```
```suggestion
`Timer::times_finished` has been renamed to `Timer::times_finished_this_tick` for clarity.
```

DJMcNab commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,401 @@
+
+### [Change gamepad.rs tuples to normal structs](https://github.com/bevyengine/bevy/pull/4519)
+
+The `Gamepad`, `GamepadButton`, `GamepadAxis`, `GamepadEvent` and `GamepadEventRaw` types are now normal structs instead of tuple structs and have a new() function.
```
I don't like the phrase "normal struct". However, there doesn't appear to be an actually given name for them.
The reference [uses](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#using-tuple-structs-without-named-fields-to-create-different-types) the term "regular struct", which isn't much better.

IceSentry replied:
I guess we can just remove "normal" and just say structs

alice-i-cecile replied:
structs with named fields?

DJMcNab commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,401 @@
+
+### [Fail to compile on 16-bit platforms](https://github.com/bevyengine/bevy/pull/4736)
+
+`bevy_ecs` will now explicitly fail to compile on 16-bit platforms, as the behavior was undefined.
```
I don't think there was any unconditional UB - only unsoundness?

IceSentry replied:
@alice-i-cecile that was your suggestion to change this. Do you have a preference for this? Otherwise I can just reuse the original test from the PR that didn't mention UB.
```suggestion
`bevy_ecs` will now explicitly fail to compile on 16-bit platforms, , due to an inability to write sound unsafe code for these platforms.
```

alice-i-cecile replied:
Yeah, I'm fine to accept that suggestion.

DJMcNab commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,401 @@
+
+### [Allow higher order systems](https://github.com/bevyengine/bevy/pull/4833)
+
+<!-- TODO marked as breaking change, but need to figure out what actually is breaking -->
```
```suggestion
`SystemParamFunction` has changed. It was not previously part of the public API, so no migration instructions are provided. (It is now included in the public API, although you still should not implement this trait for your own types).

If possible, any custom `System` implementations should be migrated to use higher order systems, which are significantly less error-prone.

Research is needed into allowing this to work for more cases.
```

DJMcNab commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,401 @@
+### [Make RenderStage::Extract run on the render world](https://github.com/bevyengine/bevy/pull/4402)
+
+The `Extract` `RenderStage` now runs on the render world (instead of the main world as before).
+You must use the `Extract` `SystemParam` to access the main world during the extract phase. `Extract` takes a single type parameter, which is any system parameter (such as `Res`, `Query` etc.). It will extract this from the main world, and returns the result of this extraction when `value` is called on it.
```
```suggestion
You must use the `Extract` `SystemParam` to access the main world during the extract phase. `Extract` takes a single type parameter, which is any system parameter (such as `Res`, `Query` etc.).
It will extract this from the main world.
Note that `Commands` will not work correctly in `Extract` - it will currently silently do nothing.
```

DJMcNab commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,401 @@
+
+// 0.8
+fn extract_clouds(mut commands: Commands, mut clouds: Extract<Query<Entity, With<Cloud>>>) {
+    for cloud in clouds.value().iter() {
```
```suggestion
    for cloud in clouds.iter() {
```

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,509 @@
+
+### [Change gamepad.rs tuples to normal structs](https://github.com/bevyengine/bevy/pull/4519)
+
+The `Gamepad`, `GamepadButton`, `GamepadAxis`, `GamepadEvent` and `GamepadEventRaw` types are now normal structs instead of tuple structs and have a new() function.
```
```suggestion
The `Gamepad`, `GamepadButton`, `GamepadAxis`, `GamepadEvent` and `GamepadEventRaw` types are now normal structs instead of tuple structs and have a `new` function.
```

rparrett commented:
I think we should expand on the `SpatialBundle` section, and place it more prominently.

It would be great to cover:

- The symptoms of needing to do this migration: stuff mysteriously not being visible
- The new components that are required for visibility propagation that are a part of the new `SpatialBundle`
- Mention that we're putting users through this because we now do a cool new thing and handle visibility hierarchies.

IceSentry commented:
@rparrett I'm pretty sure @cart will talk about that in the 0.8 blog post. I can add a quick line to explain that if things are not visible this is probably the reason.

rparrett commented:
That's a good point. Maybe something brief then. Linking off to docs for SpatialBundle or to the visibility PR might help?  But in some cases there may not be an obvious `TransformBundle` to search for to replace.

Users defining their own bundles who are not using a nested `TransformBundle`, for instance.

IceSentry commented:
@rparrett I moved it closer to the visiblity inheritance stuff and added a bit more details. If you have a better idea, leave a suggestion in a comment so that github will credit you

rparrett commented:
Actually, now that I am looking closer, `VisibilityBundle` wasn't in 0.7 so the section could use a bit more work. I'll suggest something though.

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -0,0 +1,511 @@
+}
+```
+
+### [SpatialBundle](https://github.com/bevyengine/bevy/pull/5344)
+
+If you previously needed a `TransformBundle` and a `VisibilityBundle` you should now use a `SpatialBundle`.
+
+If you were defining your own bundle with `Transform` and `Visibility` components, make sure they now have a `ComputedVisiblity` component.
````
```suggestion
### [Visibility Inheritance](https://github.com/bevyengine/bevy/pull/5310)

`Visibility` is now propagated into children in a similar way to `Transform`. Root elements of a hierarchy must now contain [`Visibility`] and [`ComputedVisiblity`] for visibility propagation to work.

[`SpatialBundle`] and [`VisibilityBundle`] have been added for convenience.

[`Visibility`]: https://docs.rs/bevy/0.8.0/bevy/render/view/struct.Visibility.html
[`ComputedVisiblity`]: https://docs.rs/bevy/0.8.0/bevy/render/view/struct.ComputedVisibility.html
[`SpatialBundle`]: https://docs.rs/bevy/0.8.0/bevy/prelude/struct.SpatialBundle.html
[`VisibilityBundle`]: https://docs.rs/bevy/0.8.0/bevy/prelude/struct.VisibilityBundle.html
```

rparrett replied:
Haha, oops.

superdump commented:
I feel like with so many items, categorisation becomes necessary. :)

IceSentry commented:
@superdump agreed, it will be done for 0.9 I already have ideas how to do it, but for 0.8 I don't have the time to categorize it manually. If anyone wants to do it I will gladly accept a PR though.

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+
+### [Visibilty Inheritance, universal ComputedVisibility and RenderLayers support](https://github.com/bevyengine/bevy/pull/5310)
+
+`Visibility` is now propagated into children in a similar way to `Transform`. Root elements of a hierarchy must now contain [`Visibility`] and [`ComputedVisiblity`] for visibility propagation to work.
```
```suggestion
[`Visibility`] is now propagated into children in a similar way to `Transform`. Root elements of a hierarchy must now contain [`Visibility`] and [`ComputedVisiblity`] for visibility propagation to work.
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+
+`Visibility` is now propagated into children in a similar way to `Transform`. Root elements of a hierarchy must now contain [`Visibility`] and [`ComputedVisiblity`] for visibility propagation to work.
+
+[`SpatialBundle`] and [`VisibilityBundle`] have been added for convenience.
+
+If you were previously reading `Visibility::is_visible` as the "actual visibility" for sprites or lights, use `ComputedVisibilty::is_visible()` instead:
```
~~~suggestion
[`SpatialBundle`] and [`VisibilityBundle`] have been added for convenience.

```rust
// 0.7
commands
    .spawn_bundle(TransformBundle::default())
    .with_children(/* ... */)

// 0.8
commands
    .spawn_bundle(SpatialBundle::default())
    .with_children(/* ... */)
```

If you were previously reading `Visibility::is_visible` as the "actual visibility" for sprites or lights, use `ComputedVisibilty::is_visible()` instead:
~~~

I'm sort of on the fence about this, so feel free to disagree/veto. The current version is just missing the simple " `TransformBundle` users can just use `SpatialBundle` now " that some users might like to see. A sentence saying something similar would take up less space.

IceSentry replied:
Since it's already pretty long I'll go with the short sentence

inodentry commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,511 @@
+
+### [Move system_param fetch struct into anonymous scope to avoid name collisions](https://github.com/bevyengine/bevy/pull/4100)
+
+For code that was using a system param's fetch struct, such as EventReader's EventReaderState, the fetch struct can now be identified via the SystemParam trait associated type Fetch, e.g. for `EventReader<T>` it can be identified as `<EventReader<'static, 'static, T>` as SystemParam>::Fetch
```
```suggestion
For code that was using a system param's fetch struct, such as EventReader's EventReaderState, the fetch struct can now be identified via the SystemParam trait associated type Fetch, e.g. for `EventReader<T>` it can be identified as `<EventReader<'static, 'static, T> as SystemParam>::Fetch`
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+### [Improve Gamepad DPad Button Detection](https://github.com/bevyengine/bevy/pull/5220)
+
+D-pad inputs can no longer be accessed as axes.
+Acess them as gamepad buttons instead.
```
```suggestion
Access them as gamepad buttons instead.
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+The trait `ReadOnlyFetch` has been replaced with `ReadOnlyWorldQuery` along with the `WorldQueryGats::ReadOnlyFetch` assoc type which has been replaced with `<WorldQuery::ReadOnly as WorldQueryGats>::Fetch`
+
+* Any where clauses such as `QueryFetch<Q>: ReadOnlyFetch` should be replaced with `Q: ReadOnlyWorldQuery`.
+* Any custom world query impls should implement `ReadOnlyWorldQuery` insead of `ReadOnlyFetch`
```
```suggestion
* Any custom world query impls should implement `ReadOnlyWorldQuery` instead of `ReadOnlyFetch`
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+fn system(query: Query<&Visibility>) {
+  for visibility in query.iter() {
+    if visibility.is_visible {
+       log!("found visible entity");
```
```suggestion
       info!("found visible entity");
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+fn system(query: Query<&ComputedVisibility>) {
+  for visibility in query.iter() {
+    if visibility.is_visible() {
+       log!("found visible entity");
```
```suggestion
       info!("found visible entity");
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+        }
+        Vec2::ZERO
+    });
+    commands.spawn().insert(ComputeVec2(task))
```
```suggestion
    commands.spawn().insert(ComputeVec2(task));
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+   query: Query<&MyComponent>,
+) {
+   query.par_for_each(&task_pool, 32, |comp| {
+        ...
```
```suggestion
        // ...
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -0,0 +1,514 @@
+```rust
+fn parallel_system(query: Query<&MyComponent>) {
+   query.par_for_each(32, |comp| {
+        ...
````
```suggestion
        // ...
```

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+    orphans: Query<(Entity, &MakeChildOf)>,
+) {
+    for (child, MakeChildOf(parent)) in &orphans {
+        commands.entity(parent).add_child(child);
```
```suggestion
        commands.entity(*parent).add_child(child);
```

rparrett reviewed (commented):
Did a quick pass on the 0.8 code blocks for compilability.

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,514 @@
+
+### [Fail to compile on 16-bit platforms](https://github.com/bevyengine/bevy/pull/4736)
+
+`bevy_ecs` will now explicitly fail to compile on 16-bit platforms, , due to an inability to write sound unsafe code for these platforms.
```
```suggestion
`bevy_ecs` will now explicitly fail to compile on 16-bit platforms, due to an inability to write sound unsafe code for these platforms.
```

rparrett commented:
@afonsolage 

The section in the migration guide for your PR https://github.com/bevyengine/bevy/pull/5210 could probably use some work, but I don't understand reflection enough to help. It's currently just using the `Changelog` section from the PR, but this doesn't really seem to give any actionable advice for a user migrating their code.

Would you mind helping out here?

https://github.com/bevyengine/bevy-website/blob/bfcc0e2538bffddf21107b96f66bc2e681c7ebf4/content/learn/book/migration-guides/0.7-0.8/_index.md#make-reflect_partial_eq-return-more-accurate-results

rparrett reviewed (commented):
Did a pass on the code blocks with the goal of formatting them consistently

rparrett commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -0,0 +1,513 @@
+
+`Camera::projection_matrix` is no longer a public field. Use the new `Camera::projection_matrix()` method instead:
+
+```rust
+
+// 0.7
+let projection = camera.projection_matrix;
````
~~~suggestion
```rust
// 0.7
let projection = camera.projection_matrix;
~~~

afonsolage commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,513 @@
+
+### [Make reflect_partial_eq return more accurate results](https://github.com/bevyengine/bevy/pull/5210)
+
+Updated [struct_trait](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates/bevy_reflect/src/struct_trait.rs#L455-L457), [tuple_struct](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates/bevy_reflect/src/tuple_struct.rs#L366-L368), [tuple](https://github.com/bevyengine/bevy/blob/dfe969005264fff54060f9fb148639f80f9cfb29/crates ...
```
Previously all reflect `***_partial_eq` helper methods returned `Some(false)` even when the comparison could not be performed, which was misleading. Now those helper methods will return `None` if the comparison could not be performance, while still returning either `Some(true)` or `Some(false)` based on the equality comparison.

IceSentry replied:
How about this:
```markdown
Previously, all reflect `***_partial_eq` helper methods returned `Some(false)` when the comparison could not be performed, which was misleading. They now return `None` when the comparison cannot be performed.
```

rparrett replied:
That seems like a huge improvement and the English is perfect. Thanks!

IceSentry replied:
If you like it, could you make a code suggestion with the changes. That way you will be credited as a co-author.

afonsolage replied:
You deserve to be the co-author, this summary is way cooler 😄. I'm glad to had whenever I can

afonsolage reviewed (commented):
@rparrett sure, I'm not native, so I may not find the correct words to have a good migration guide, but here is what this PR does with more details

aevyrie commented:
Added some content in a PR: https://github.com/IceSentry/bevy-website/pull/4

ndarilek commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
````diff
@@ -0,0 +1,508 @@
+```rs
+// 0.7
+fn system(mut commands: Commands) {
+    let task = thread_pool.spawn(async move {
````
For completeness' sake, do we want to show where `thread_pool` might come from in this example?

IceSentry replied:
Yes, you're right, I completely forgot to do that

ndarilek commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,508 @@
+struct ComputeVec2(Task<Vec2>);
+
+fn system(mut commands: Commands) {
+    let task = thread_pool.spawn(async move {
```
Same here as well?

alice-i-cecile commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,522 @@
+
+`bevy_asset` and `bevy_scene` are no longer enabled when `no-default-features` is used with the `bevy` dependency.
+
+* Crates that use bevy with `no-default-features` will need to add these features manually
```
```suggestion
* Crates that use Bevy with `no-default-features` will need to add these features manually if they rely on them.
```

komadori commented on content/learn/book/migration-guides/0.7-0.8/_index.md:
```diff
@@ -0,0 +1,522 @@
+Renamed `HasRawWindowHandleWrapper` to `ThreadLockedRawWindowHandleWrapper`.
+
+### [Migrate to encase from crevice](https://github.com/bevyengine/bevy/pull/4339)
+
```
#### AsStd140, AsStd430 -> ShaderType
* Replaced as_std140(), as_std430() with
```
let byte_buffer = [0u8; YourDataType::SIZE.get() as usize];
let mut buffer = encase::UniformBuffer::new(byte_buffer);
buffer.write(&value).unwrap();```

cart reviewed (approved):
Great work everyone!

cart commented:
bors r+
